includes/abstracts/abstract-wc-data.php

```diff
@@ -0,0 +1,331 @@
+			$cached_meta = wp_cache_get( $cache_key, $this->_cache_group );
+
+			if ( false !== $cached_meta ) {
+				$this->_meta_data = $cached_meta;
+				$cache_loaded = true;
+			}
+		}
+
+		if ( ! $cache_loaded ) {
+			global $wpdb;
+			$db_info = $this->_get_db_info();
+			$raw_meta_data = $wpdb->get_results( $wpdb->prepare( "
+				SELECT " . $db_info['meta_id_field'] . ", meta_key, meta_value
+				FROM " . $db_info['table'] . "
+				WHERE " . $db_info['object_id_field'] . " = %d ORDER BY " . $db_info['meta_id_field'] . "
+			", $this->get_id() ) );
+
+			foreach ( $raw_meta_data as $meta ) {
+				if ( in_array( $meta->meta_key, $this->get_internal_meta_keys() ) ) {
+					continue;
+				}
+
+				$this->_meta_data[ $meta->{$db_info['meta_id_field']} ] = (object) array( 'key' => $meta->meta_key, 'value' => $meta->meta_value );
+			}
+
+			if ( ! empty ( $this->_cache_group ) ) {
+				wp_cache_set( $cache_key, $this->_meta_data, $this->_cache_group );
+			}
+		}
+	}
+
+	/**
+	 * Update Meta Data in the database.
+	 * @since 2.7.0
+	 */
+	protected function save_meta_data() {
+		global $wpdb;
+		$db_info = $this->_get_db_info();
+		$all_meta_ids = array_map( 'absint', $wpdb->get_col( $wpdb->prepare( "
+			SELECT " . $db_info['meta_id_field'] . " FROM " . $db_info['table'] . "
+			WHERE " . $db_info['object_id_field'] . " = %d", $this->get_id() ) . "
+			AND meta_key NOT IN ('" . implode( "','", array_map( 'esc_sql', $this->get_internal_meta_keys() ) ) . "');
+		" ) );
+		$set_meta_ids = array();
+
+		foreach ( $this->_meta_data as $meta_id => $meta ) {
+			if ( 'new' === substr( $meta_id, 0, 3 ) ) {
+				$set_meta_ids[] = add_metadata( $this->_meta_type, $this->get_id(), $meta->key, $meta->value, false );
+			} else {
+				update_metadata_by_mid( $this->_meta_type, $meta_id, $meta->value, $meta->key );
+				$set_meta_ids[] = absint( $meta_id );
+			}
+		}
+
+		// Delete no longer set meta data
+		$delete_meta_ids = array_diff( $all_meta_ids, $set_meta_ids );
+
+		foreach ( $delete_meta_ids as $meta_id ) {
+			delete_metadata_by_mid( $this->_meta_type, $meta_id );
+		}
+
+		if ( ! empty ( $this->_cache_group ) ) {
+			WC_Cache_Helper::incr_cache_prefix( $this->_cache_group );
+		}
+		$this->read_meta_data();
+	}
+
+	/**
+	 * Table structure is slightly different between meta types, this function will return what we need to know.
+	 * @since 2.7.0
+	 * @return array Array elements: table, object_id_field, meta_id_field
+	 */
+	private function _get_db_info() {
+		global $wpdb;
+
+		$object_id_field = '';
+		$meta_id_field   = 'meta_id'; // for some reason users calls this umeta_id so we need to track this as well.
+		$table           = $wpdb->prefix;
+		// If we are dealing with a type of metadata that is not a core type, the table should be prefixed.
+		if ( ! in_array( $this->_meta_type, array( 'post', 'user', 'comment', 'term' ) ) ) {
+			$table .= 'woocommerce_';
```

Comment:
This statement could also be used for obiect_id_field if you wanted to cut down the size of this function. Each obiect_id_field is named {$this->_meta_type}_id to save on some code.